be/src/formats/parquet/scalar_column_reader.cpp

```diff
@@ -399,32 +385,111 @@ bool ScalarColumnReader::try_to_use_dict_filter(ExprContext* ctx, bool is_decode
+                    TypeDescriptor::from_logical_type(ColumnDictFilterContext::kDictCodePrimitiveType), true);
+        }
+        _ori_column = dst;
+        dst = _tmp_code_column;
+        dst->reserve(range.span_size());
+        SCOPED_RAW_TIMER(&_opts.stats->column_read_ns);
+        return _reader->read_range(range, filter, content_type, dst.get());
+    } else {
+        if (!_converter->need_convert) {
+            SCOPED_RAW_TIMER(&_opts.stats->column_read_ns);
+            return _reader->read_range(range, filter, content_type, dst.get());
         } else {
-            dst->append_default(src->size());
+            if (_tmp_intermediate_column == nullptr) {
+                _tmp_intermediate_column = _converter->create_src_column();
+            }
+            _tmp_intermediate_column->reserve(range.span_size());
+            {
+                SCOPED_RAW_TIMER(&_opts.stats->column_read_ns);
+                RETURN_IF_ERROR(_reader->read_range(range, filter, content_type, _tmp_intermediate_column.get()));
+            }
+            if constexpr (!LAZY_CONVERT) {
+                {
+                    SCOPED_RAW_TIMER(&_opts.stats->column_convert_ns);
+                    RETURN_IF_ERROR(_converter->convert(_tmp_intermediate_column, dst.get()));
+                }
+                _tmp_intermediate_column->reset_column();
+                return Status::OK();
+            } else {
+                _ori_column = dst;
+                dst = _tmp_intermediate_column;
+                return Status::OK();
+            }
         }
+    }
+}
+
+Status ScalarColumnReader::_dict_decode(ColumnPtr& dst, ColumnPtr& src) {
+    Column* dict_values = ColumnHelper::get_data_column(dst.get());
+    dict_values->reserve(src->size());
+
+    // decode dict code to dict values.
+    // note that in dict code, there could be null value.
+    const ColumnPtr& dict_codes = src;
+    auto* codes_nullable_column = ColumnHelper::as_raw_column<NullableColumn>(dict_codes);
+    auto* codes_column = ColumnHelper::as_raw_column<FixedLengthColumn<int32_t>>(codes_nullable_column->data_column());
+    RETURN_IF_ERROR(_reader->get_dict_values(codes_column->get_data(), *codes_nullable_column, dict_values));
+    DCHECK_EQ(dict_codes->size(), dict_values->size());
+    if (dst->is_nullable()) {
+        auto* nullable_values = down_cast<NullableColumn*>(dst.get());
+        nullable_values->swap_null_column(*codes_nullable_column);
+    }
+    src->reset_column();
+    return Status::OK();
+}
 
-        src->reset_column();
+template <bool LAZY_DECODE, bool LAZY_CONVERT>
+Status ScalarColumnReader::_fill_dst_column_impl(ColumnPtr& dst, ColumnPtr& src) {
+    if constexpr (LAZY_DECODE) {
+        if (_dict_filter_ctx != nullptr && !_dict_filter_ctx->is_decode_needed) {
+            dst->append_default(src->size());
+            src->reset_column();
+        } else {
+            if constexpr (LAZY_CONVERT) {
```

Comment:
Can `LAZY_DECODE` and `LAZY_CONVERT` both be true at the same time?

Reply:
no, LAZY_DECODE is only used for dictionary encoded string, it's impossible for this type has a converter.